src/group_theory/nilpotent.lean

```diff
@@ -0,0 +1,290 @@
+    series* if it starts at `G` and `⁅H n, ⊤⁆ ⊆ H (n + 1)` for all `n`. Note that this series
+    may not terminate at `⊥`, and the `H i` need not be normal. Similarly a sequence is an
+    *ascending central series* if `H 0 = ⊥` and `⁅H (n + 1), ⊤⁆ ⊆ H n` for all `n`, again with no
+    requirement that the series reaches `⊤` or that the `H i` are normal.
+
+## Main theorems
+
+`G` is *defined* to be nilpotent if the upper central series reaches `⊤`.
+* `nilpotent_iff_finite_ascending_central_series` : `G` is nilpotent iff some ascending central
+    series reaches `⊤`.
+* `nilpotent_iff_finite_descending_central_series` : `G` is nilpotent iff some descending central
+    series reaches `⊥`.
+* `nilpotent_iff_lower` : `G` is nilpotent iff the lower central series reaches `⊥`.
+
+## Warning
+
+A "central series" is usually defined to be a finite sequence of normal subgroups going
+from `⊥` to `⊤` with the property that each subquotient is contained within the centre of
+the associated quotient of `G`. This means that if `G` is not nilpotent, then
+none of what we have called `upper_central_series G`, `lower_central_series G` or
+the sequences satisfying `is_ascending_central_series` or `is_descending_central_series`
+are actually central series. Note that the fact that the upper and lower central series
+are not central series if `G` is not nilpotent is a standard abuse of notation.
+
+-/
+
+open subgroup
+
+variables {G : Type*} [group G] (H : subgroup G) [normal H]
+
+/-- If `H` is a normal subgroup of `G`, then the set `{x : G | ∀ y : G, x*y*x⁻¹*y⁻¹ ∈ H}`
+is a subgroup of `G` (because it is the preimage in `G` of the centre of the
+quotient group `G/H`.)
+-/
+def upper_central_series_step : subgroup G :=
+{ carrier := {x : G | ∀ y : G, x * y * x⁻¹ * y⁻¹ ∈ H},
+  one_mem' := λ y, by simp [subgroup.one_mem],
+  mul_mem' := λ a b ha hb y, begin
+    convert subgroup.mul_mem _ (ha (b * y * b⁻¹)) (hb y) using 1,
+    group,
+  end,
+  inv_mem' := λ x hx y, begin
+    specialize hx y⁻¹,
+    rw [mul_assoc, inv_inv] at ⊢ hx,
+    exact subgroup.normal.mem_comm infer_instance hx,
+  end }
+
+open quotient_group
+
+/-- The proof that `upper_central_series_step H` is the preimage of the centre of `G/H` under
+the canonical surjection. -/
+lemma upper_central_series_step_eq_comap_center :
+  upper_central_series_step H = subgroup.comap (mk' H) (center (quotient H)) :=
+begin
+  ext,
+  rw [mem_comap, mem_center_iff],
+  change (∀ y, x * y * x⁻¹ * y⁻¹ ∈ H) ↔ _,
+  split,
+  { intros h q,
+    apply induction_on q,
+    intro y,
+    change ((y * x : G) : quotient H) = (x * y : G),
+    rw [eq_comm, eq_iff_div_mem, div_eq_mul_inv],
+    convert h y using 1, group,
+  },
+  { intros h y,
+    specialize h y,
+    change ((y * x : G) : quotient H) = (x * y : G) at h,
+    rw [eq_comm, eq_iff_div_mem, div_eq_mul_inv] at h,
+    convert h using 1, group },
+end
```

Comment:
Here's a shorter proof:
```suggestion
lemma quotient_group.forall (H : subgroup G) {C : quotient_group.quotient H → Prop} :
  (∀ x : quotient_group.quotient H, C x) ↔ ∀ x : G, C x :=
⟨λ hx x, hx _, quot.ind⟩

/-- The proof that `upper_central_series_step H` is the preimage of the centre of `G/H` under
the canonical surjection. -/
lemma upper_central_series_step_eq_comap_center :
  upper_central_series_step H = subgroup.comap (mk' H) (center (quotient H)) :=
begin
  ext,
  rw [mem_comap, mem_center_iff, quotient_group.forall],
  apply forall_congr,
  intro y,
  change x * y * x⁻¹ * y⁻¹ ∈ H ↔ ((y * x : G) : quotient H) = (x * y : G),
  rw [eq_comm, eq_iff_div_mem, div_eq_mul_inv],
  congr' 2,
  group,
end
```
`quotient_group.forall` obviously belongs somewhere else

Reply:
Oh this is very nice! Yes, I didn't like my proof -- too much repetition. I'll update.

Reply:
I put my proof in group_theory.coset. Note that apparently you can't call a lemma `forall` :-)

Reply:
You absolutely can call it forall, you just have to either write `protected lemma the_namespace.forall` or if you're already in the namespace, `protected lemma «forall»`